lib/gingr/watcher.rb

```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
+require 'listen'
+require 'open3'
+require_relative 'config'
+
+module Gingr
+  class Watcher
+    include Config
+
+    # Only watch for files in WATCHED_DIRECTORIES[:READY] that match this pattern
+    WATCH_FILTER = Regexp.compile(/\.zip$/)
+
+    WATCHED_DIRECTORIES = {
+      # Directory into which new files are dropped when ready for processing.
+      # .watch! monitors this for new .zip's.
+      READY: 'ready'.freeze,
+      # Directory into which successfully processed files are moved post-processing.
+      PROCESSED: 'processed'.freeze,
+      # Directory into which failed files are moved post-processing.
+      FAILED: 'failed'.freeze,
+    }
+
+    attr_reader :options
+    attr_reader :root_dir
+
+    def initialize(root_dir, *options)
+      @root_dir = root_dir
+      @options = options
+
+      validate_directories!
+    end
+
+    def start!
+      start
+      sleep
+    end
+
+    def start
+      Config.logger.info("Monitoring directory for new zipfiles: #{ready_dir}")
+      listener.start unless listener.processing?
+    end
+
+    def listener
+      @listener ||= begin
+        Listen.to(ready_dir, only: WATCH_FILTER, force_polling: true) do |_, added, _|
+          added.each do |zipfile|
+            Config.logger.debug("Processing zipfile: #{zipfile}")
+            exec_gingr_with_file!(zipfile)
+          end
+        end
+      end
+    end
+
+    def ready_dir
+      @ready_dir ||= File.join(@root_dir, WATCHED_DIRECTORIES[:READY])
+    end
+
+    def processed_dir
+      @processed_dir ||= File.join(@root_dir, WATCHED_DIRECTORIES[:PROCESSED])
+    end
+
+    def failed_dir
+      @failed_dir ||= File.join(@root_dir, WATCHED_DIRECTORIES[:FAILED])
+    end
+
+    def exec_gingr_with_file!(zipfile)
+      begin
+        stdout, stderr, status = Open3.capture3('gingr', 'all', zipfile, *options)
```

Comment:
It would be preferable to call the code directly via Ruby, but that will require refactoring the existing code.

Reply:
It makes sense, sure